### ISSUE TITLE
IM6 -> IM7 MagickSimilarityImage

### ISSUE THREAD

**Issue by Danack** [ImageMagick/ImageMagick#35]
The value set for similarity in the function MagickSimilarityImage has changed dramatically. Using the same code and image, in IM6 the value is about 0.559; for IM7 it is 390. Is that a deliberate change or a regression?

I can provide a C code example if required, but it's showing that difference in the PHP Imagick tests.

**Comment by dlemstra** [comment]
IM7 now also includes the alpha channel. That is probably where the difference is coming from. And did you use the same similarity_threshold?

**Comment by Danack** [comment]
> And did you use the same similarity_threshold?

Yes - I found the issue from the result of the PHP Imagick tests. https://travis-ci.org/mkoppanen/imagick/builds/81971577 The same code is run for IM6 and IM7. 

I've posted the PHP code below, with an added line to deactivate the alpha channel. That didn't seem to make any difference.

The only difference between the images is made by the call to vignetteImage which shouldn't affect the alpha channel should it?

The location and size of the similarityImage is correct for the test. It's only the score that's changed.


```
<?php

$imagick = new \Imagick();
$imagick->newPseudoImage(640, 480, "magick:logo");
$imagick->adaptiveResizeImage(100, 100, true);
$imagick->setImageAlphaChannel(Imagick::ALPHACHANNEL_DEACTIVATE);

$imagick2 = clone $imagick;
$imagick2->cropimage(25, 25, 50, 50);
$imagick2->setImageFormat('png');

$imagick2->setImagePage(25, 25, 0, 0);
$imagick2->vignetteimage(0.25, 0.75, 3, 3);

$similarity = 'not set';
$bestMatch = 'not set';
$comparison = $imagick->subImageMatch($imagick2, $bestMatch, $similarity);

$comparison->setImageFormat('png');
$bytes = $imagick->getImageBlob();
if (strlen($bytes) <= 0) { echo "Failed to generate image.";} 

if ($similarity >= 0.558 && $similarity <= 0.559) {
	echo "Similarity in range\n";
}
else {
	echo "Similarity is out of range.\n";
	var_dump($similarity);
}

ksort($bestMatch);
foreach ($bestMatch as $key => $value) {
	echo "$key : $value\n";
}

```

**Comment by urban-warrior** [comment]
Can you provide C code or a command-line example?  Or a link where to download imagick that can compile against ImageMagick version 7?  We ran your PHP example against IMv6 but imagick-3.3.0RC2 failed to compile against IMv7.

**Comment by urban-warrior** [comment]
> I can provide a C code example if required

Yes please.  Or a command line example.  We need to reproduce the problem before we can comment further.

**Comment by dlemstra** [comment]
@Danack Could you please respond to our question?

**Comment by Danack** [comment]
Sorry for the delay. Testing against master of 7 and 6.9.2-5 shows a difference in the output:

```
For IM7 compiled from master output is:
Similarity is 390.000000

For ImageMagick-6.9.2-5 output is:
Similarity is 0.558027
```

Code is:
```
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

// comment in this for 7.
//#define IM7 

#ifdef IM7
#include <MagickWand/MagickWand.h>
#else
#include <wand/MagickWand.h>
#endif

int main(int argc,char **argv) {

    MagickWand *small_sample_wand;
    MagickWand *original_wand;
    MagickWand *comparison_wand;
    
    MagickWand *tmp_wand;
    int c;
    int x = 48;
    char filename[1024];
    
    MetricType metric = AbsoluteErrorMetric;
    double similarity_threshold = 1.0;

    RectangleInfo best_match_offset;
    double similarity = -1;

    MagickWandGenesis();
    
    original_wand = NewMagickWand();
    
    //$imagick->newPseudoImage(640, 480, "magick:logo");
    MagickSetSize(original_wand, 640, 480);
    MagickReadImage(original_wand, "magick:logo");

    //$imagick->adaptiveResizeImage(100, 100, true);
    MagickAdaptiveResizeImage(original_wand, 100, 75);
    //$imagick->setImageAlphaChannel(Imagick::ALPHACHANNEL_DEACTIVATE);
    MagickSetImageAlphaChannel(original_wand, DeactivateAlphaChannel);

    small_sample_wand = CloneMagickWand(original_wand);

    MagickCropImage(small_sample_wand, 25, 25, 50, 50);
    MagickSetImageFormat(small_sample_wand, "png");
    MagickSetImagePage(small_sample_wand, 25, 25, 0, 0);
    MagickVignetteImage(small_sample_wand, 0.25, 0.75, 3, 3);

    comparison_wand = MagickSimilarityImage(
        original_wand,
        small_sample_wand,
#ifdef IM7
        metric,
        similarity_threshold,
#endif

        &best_match_offset, 
        &similarity
    );
 
    printf("Similarity is %f \n", similarity);

    //For IM7 compiled from master output is:
    //Similarity is 390.000000
    //For ImageMagick-6.9.2-5 output is:
    //Similarity is 0.558027

    MagickWandTerminus();
    return (0);
}

// Both were compiled with:
//
// ./configure --with-quantum-depth=16 \
//           --with-magick-plus-plus=no \
//           --without-perl \
//           --disable-static \
//           --exec-prefix=/usr/sbin \
//           --disable-hdri \
//           --disable-docs \
//           --disable-openmp 

```

**Comment by urban-warrior** [comment]
Write the image to disk and compare with

  compare -metric rmse ~/imv7.png  imv6.png x:
  5.14 (7.84314e-05)  

We're seeing a one pixel difference in the green component.  With such a low distortion, we're not concerned about this difference.

**Comment by Danack** [comment]
Hmm. I did that and I think the results look wrong.

IM 6 image
![im35_6](https://cloud.githubusercontent.com/assets/1505719/11047946/f57c1844-872c-11e5-9be9-3e2219569378.png)

IM7 image - mostly transparent
![im35_7](https://cloud.githubusercontent.com/assets/1505719/11047945/f579e240-872c-11e5-8f1e-4c843566a0fe.png) https://cloud.githubusercontent.com/assets/1505719/11047945/f579e240-872c-11e5-8f1e-4c843566a0fe.png

Additionally running the executable of the code above for IM7 through valgrind gives several warnings about uninitialized variables.

```
Conditional jump or move depends on uninitialised value(s)
==10082==    at 0x5437900: WriteOnePNGImage (png.c:8568)
==10082==    by 0x544206F: WritePNGImage (png.c:12159)
==10082==    by 0x5200CBA: WriteImage (constitute.c:1091)
==10082==    by 0x4EC77C0: MagickWriteImage (magick-image.c:12410)
==10082==    by 0x400C5A: main (in /home/github/imagemagicktest/ImageMagick35)
```

I'm unable to run the compare program on IM7 as is gives a cryptic error message:

```
# /usr/sbin/bin/compare -metric rmse ./IM35_6.png ./IM35_7.png 
compare:  `./IM35_6.png' @ error/compare.c/CompareImagesCommand/937.
```

This is all on Centos 6.4 x86_64

**Comment by urban-warrior** [comment]
ImageMagick version 6 only supports the RootMeanSquaredErrorMetric metric.  If you want similar results from IMv6 and IMv7, set the metric for IMv7 to RootMeanSquaredErrorMetric and for IMv6 you need to define the similarity-threshold to 1.0 with

    MagickSetImageArtifact(original_wand,"compare:similarity-threshold","1.0");

With these changes, we get a similarity of 0.764258 for both IMv6 and IMv7.  If you want to use a metric other than RootMeanSquaredErrorMetric in IMv6, you'll need to call MagickCore's SimilarityMetricImage() method which has a metric argument.

Ideally we would call SimilarityMetricImage() from MagickWand's MagickSimilarityImage() but we cannot break the API / ABI by changing its signature.

**Comment by Danack** [comment]
Ok, so I was using the wrong metric. I should be using RootMeanSquaredErrorMetric in 7 to have a similar comparison. I will open a separate issue for the output of the image as it might be a separate issue.

However I cannot replicate getting a similar number for the similarity score.

Additionally, I have just realised that the position reported seems bogus for IM7. It is reporting a best match at position 0, 0 when the best match should be at 50,50.

Updated code with the position being output.

```
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define IM7 

#ifdef IM7
#include <MagickWand/MagickWand.h>
#else
#include <wand/MagickWand.h>
#endif
 
int main(int argc,char **argv) {

    MagickWand *small_sample_wand;
    MagickWand *original_wand;
    MagickWand *comparison_wand;
    
    MagickWand *tmp_wand;
    int c;
    int x = 48;
    char filename[1024];
    
    MetricType metric = RootMeanSquaredErrorMetric;
    double similarity_threshold = 1.0;

    RectangleInfo best_match_offset;
    double similarity = -1;

    MagickWandGenesis(); 
    
    original_wand = NewMagickWand();
    
    MagickSetSize(original_wand, 640, 480);
    MagickReadImage(original_wand, "magick:logo");
    MagickAdaptiveResizeImage(original_wand, 100, 75);
    MagickSetImageAlphaChannel(original_wand, DeactivateAlphaChannel);

    small_sample_wand = CloneMagickWand(original_wand);

    MagickCropImage(small_sample_wand, 25, 25, 50, 50);
    MagickSetImageFormat(small_sample_wand, "png");
    MagickSetImagePage(small_sample_wand, 25, 25, 0, 0);
    MagickVignetteImage(small_sample_wand, 0.25, 0.75, 3, 3);

    comparison_wand = MagickSimilarityImage(
        original_wand,
        small_sample_wand,
#ifdef IM7
        metric,
        similarity_threshold,
#endif
        &best_match_offset, 
        &similarity
    );
 
    printf("Similarity is %f \n", similarity);

    MagickSetImageFormat(comparison_wand, "png");

#ifdef IM7
    MagickWriteImage(comparison_wand, "./IM35_7.png");
#else
    MagickWriteImage(comparison_wand, "./IM35_6.png");
#endif

    printf(
        "Width: %d Height: %d x: %d, y: %d\n",
        best_match_offset.width,
        best_match_offset.height,
        best_match_offset.x,
        best_match_offset.y
    );


    //For IM7 compiled from master output is:
    // Similarity is 0.764258 
    // Width: 25 Height: 25 x: 0, y: 0
    //
    // For IM6 compiled from head of the IM6 branch 
    // Similarity is 0.558027 
    // Width: 25 Height: 25 x: 50, y: 50

    

    MagickWandTerminus();
    return (0);
}
```

```
    For IM7 compiled from master output is:
    Similarity is 0.764258 
    Width: 25 Height: 25 x: 0, y: 0
    
    For IM6 compiled from head of the IM6 branch 
    Similarity is 0.558027 
    Width: 25 Height: 25 x: 50, y: 50
```


```
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define IM7 

#ifdef IM7
#include <MagickWand/MagickWand.h>
#else
#include <wand/MagickWand.h>
#endif
 
int main(int argc,char **argv) {

    MagickWand *small_sample_wand;
    MagickWand *original_wand;
    MagickWand *comparison_wand;
    
    MagickWand *tmp_wand;
    int c;
    int x = 48;
    char filename[1024];
    
    MetricType metric = RootMeanSquaredErrorMetric;
    double similarity_threshold = 1.0;

    RectangleInfo best_match_offset;
    double similarity = -1;

    MagickWandGenesis(); 
    
    original_wand = NewMagickWand();
    
    MagickSetSize(original_wand, 640, 480);
    MagickReadImage(original_wand, "magick:logo");
    MagickAdaptiveResizeImage(original_wand, 100, 75);
    MagickSetImageAlphaChannel(original_wand, DeactivateAlphaChannel);

    small_sample_wand = CloneMagickWand(original_wand);

    MagickCropImage(small_sample_wand, 25, 25, 50, 50);
    MagickSetImageFormat(small_sample_wand, "png");
    MagickSetImagePage(small_sample_wand, 25, 25, 0, 0);
    MagickVignetteImage(small_sample_wand, 0.25, 0.75, 3, 3);

    comparison_wand = MagickSimilarityImage(
        original_wand,
        small_sample_wand,
#ifdef IM7
        metric,
        similarity_threshold,
#endif
        &best_match_offset, 
        &similarity
    );
 
    printf("Similarity is %f \n", similarity);

    MagickSetImageFormat(comparison_wand, "png");

#ifdef IM7
    MagickWriteImage(comparison_wand, "./IM35_7.png");
#else
    MagickWriteImage(comparison_wand, "./IM35_6.png");
#endif

    printf(
        "Width: %d Height: %d x: %d, y: %d\n",
        best_match_offset.width,
        best_match_offset.height,
        best_match_offset.x,
        best_match_offset.y
    );


    //For IM7 compiled from master output is:
    // Similarity is 0.764258 
    // Width: 25 Height: 25 x: 0, y: 0
    //
    // For IM6 compiled from head of the IM6 branch 
    // Similarity is 0.558027 
    // Width: 25 Height: 25 x: 50, y: 50

    

    MagickWandTerminus();
    return (0);
}

// Both were compiled with:
//
// ./configure --with-quantum-depth=16 \
//           --with-magick-plus-plus=no \
//           --without-perl \
//           --disable-static \
//           --exec-prefix=/usr/sbin \
//           --disable-hdri \
//           --disable-docs \
//           --disable-openmp 


```

**Comment by mikayla-grace** [comment]
A difference in similarity is expected.  The algorithms in IMv7, in some cases, behave differently than those in IMv6.  If you see a significant difference for a particular command-line option, let us know-- it could be a bug.  Minor differences are expected and generally not worthy of concern.

We modified your code slightly to read the rose: image and a sharpened rose image and both IMv6 and IMv7 returned the same similarity metric of 0.043868.   Try applying one image-processing algorithm to your source image and compare the similarity metric between IMv6 and IMv7.  If it has a significant difference, let us know and we will determine if there is a bug in IMv7.

**Comment by Danack** [comment]
Ok, the difference in similarity is fine. I will open a separate issue for the wrong location as it seems to have caused confusion having it in the same issue.
